### PR TITLE
feat: Retry the token import when starting offline

### PR DESCRIPTION
**Pull request by plmercereau**
On startup, retries to import the token until network is up and the server does not respond a 5xx error: the 5 first times every second, then every 5 second.

In the meantime the authentication status is loading.

`nhost.auth.getAuthenticationStatus()` and the `useAuthenticationStatus` react hoos/vue composable get a new `connectionAttempts` property.

### CODE REVIEW

File: packages/core/src/machines/index.ts [hunk 1]
```diff
@@ -137,7 +142,20 @@ export const createAuthMachine = ({
                   actions: ['saveSession', 'reportTokenChanged'],
                   target: 'signedIn'
                 },
-                onError: { actions: ['saveAuthenticationError'], target: 'signedOut' }
+                onError: [
```
Comment (szilarddoro) [inline packages/core/src/machines/index.ts:145]: Could you please write tests for this?

**Review by szilarddoro — CHANGES_REQUESTED** [review]

File: packages/core/src/machines/index.ts (hunk 1 above)
Comment (plmercereau) [reply packages/core/src/machines/index.ts:145]: Yes, as soon as we will have validated the design

**Comment by elitan** [comment]
Sounds good to me.

Just a few questions: 

1. What are the values of `isLoading` and `isAuthenticated` after the 20 tries?
2. What happens if Auth returns e.g. 500 (instead of a network error)?

**Comment by plmercereau** [comment]
> 1. What are the values of `isLoading` and `isAuthenticated` after the 20 tries?

after the attempts, `isLoading: false` and `isAuthenticated: true`

> 2. What happens if Auth returns e.g. 500 (instead of a network error)?

It will not retry and will return `isAuthenticated: false`

**Comment by elitan** [comment]
Understood.

What do you think about trying to authenticate indefinitely? This way, we would never get into an `isLoading: false, isAuthenticated: false` state when the user might have a valid refresh token.

Setting `isAuthenticated: false` when the server returns `500` also feels wrong to me. Again, it could be that the user has a valid refresh token. 

Going back to what I wrote [originally](https://github.com/nhost/nhost/issues/752#issue-1284818107), I think this is how it should work:

1. `isLoading` should only return `false` once the SDK is 100% sure about the `isAuthenticated` state. That means that `isLoading` should only be set to `false` if the SDK gets an actual response from `/v1/auth/token`. Either a `200 OK` or `401 Unauthorize`.
2. If the SDK can't reach `/v1/auth/token` because the network connection is unstable, the SDK should retry the request every 5 seconds until it successfully gets a response.
3. (not sure about this one): We can also introduce a new variable to `useAuthenticationStatus` indicating the reason for `isLoading` to be `false`. The reason will always be because the SDK is waiting for a network request, but it could be helpful to inform whether the network is stable and the SDK is making the first request or if the network is unstable and the SDK is making the 10th request.

**Comment by plmercereau** [comment]
then we could add `connectionAttempts` to let the developer decide what to do after x attempts.
I would keep the first five attempts's delay to 1 second and the following to 5 seconds - 5 seconds is very long to wait in a unreliable internet connection that only failed to get one call

**Comment by plmercereau** [comment]
but still, `nhost.auth.signIn` will never resolve when offline

File: packages/core/src/machines/index.ts [hunk 2]
```diff
@@ -869,11 +890,22 @@ export const createAuthMachine = ({
               })
               return { session, error: null }
             } catch (exception) {
-              error = (exception as { error: ValidationErrorPayload }).error
+              error = (exception as { error: ErrorPayload }).error
```
Comment (szilarddoro) [inline packages/core/src/machines/index.ts:893]: You can return the rejected promise when this error happens

**Review by szilarddoro — APPROVED** [review]
Looks good! 👍

**Pull request merged**